File: tests/typesafety/typescript/io_actions.ts
// expect: passes

class ConnectionLost extends Error {}

type IO<T> = { readonly run: () => T };

function of<T>(value: T): IO<T> {
  return { run: () => value };
}

function andThen<T, U>(action: IO<T>, step: (value: T) => IO<U>): IO<U> {
  return { run: () => step(action.run()).run() };
}

function retry<T>(times: number, action: IO<T>): IO<T> {
  return {
    run: () => {
      let failures = 0;
      for (;;) {
        try {
          return action.run();
        } catch (error) {
          if (!(error instanceof ConnectionLost)) throw error;
          failures += 1;
          if (failures >= times) throw error;
        }
      }
    },
  };
}

// A connection that fails twice and then answers, so the retry is observable.
let calls = 0;

function flakyFetch(): string {
  calls += 1;
  if (calls < 3) throw new ConnectionLost("try again");
  return "payload";
}

export const greeting = andThen(retry(3, { run: flakyFetch }), (text) => of(text.toUpperCase()));
// Nothing has run yet; greeting.run() answers "PAYLOAD" on the third call.
// The Python twin runs these assertions in CI.

File: tests/typesafety/python/io_actions.py
# expect: passes
# run: yes
# title: IO describes the fetch, retry wraps it, and run performs it
# improves: inline_retry
from collections.abc import Callable
from dataclasses import dataclass


@dataclass(frozen=True)
class IO[T]:
    run: Callable[[], T]


def of[T](value: T) -> IO[T]:
    return IO(lambda: value)


def and_then[T, U](action: IO[T], step: Callable[[T], IO[U]]) -> IO[U]:
    return IO(lambda: step(action.run()).run())


def retry[T](times: int, action: IO[T]) -> IO[T]:
    def attempt() -> T:
        failures = 0
        while True:
            try:
                return action.run()
            except ConnectionError:
                failures += 1
                if failures >= times:
                    raise
    return IO(attempt)


# A connection that fails twice and then answers, so the retry is observable.
_calls = {"count": 0}


def flaky_fetch() -> str:
    _calls["count"] += 1
    if _calls["count"] < 3:
        raise ConnectionError("try again")
    return "payload"


greeting = and_then(retry(3, IO(flaky_fetch)), lambda text: of(text.upper()))

assert _calls["count"] == 0  # nothing has run yet
assert greeting.run() == "PAYLOAD"
assert _calls["count"] == 3  # two failures, one answer
